crates/sui-types/src/messages_checkpoint.rs

```diff
@@ -188,6 +188,171 @@ pub struct CheckpointSummary {
+        }
+
+        let mut this = Self::default();
+        if left.epoch != right.epoch {
+            this.epoch = Some((left.epoch, right.epoch));
+        }
+        if left.sequence_number != right.sequence_number {
+            this.sequence_number = Some((left.sequence_number, right.sequence_number));
+        }
+        if left.network_total_transactions != right.network_total_transactions {
+            this.network_total_transactions = Some((
+                left.network_total_transactions,
+                right.network_total_transactions,
+            ));
+        }
+        if left.content_digest != right.content_digest {
+            this.content_digest = Some((left.content_digest, right.content_digest));
+        }
+        if left.previous_digest != right.previous_digest {
+            this.previous_digest = Some((left.previous_digest, right.previous_digest));
+        }
+        if left.epoch_rolling_gas_cost_summary != right.epoch_rolling_gas_cost_summary {
+            this.epoch_rolling_gas_cost_summary = Some((
+                left.epoch_rolling_gas_cost_summary.clone(),
+                right.epoch_rolling_gas_cost_summary.clone(),
+            ));
+        }
+        if left.timestamp_ms != right.timestamp_ms {
+            this.timestamp_ms = Some((left.timestamp_ms, right.timestamp_ms));
+        }
+        if left.checkpoint_commitments != right.checkpoint_commitments {
+            this.checkpoint_commitments = Some((
+                left.checkpoint_commitments.clone(),
+                right.checkpoint_commitments.clone(),
+            ));
+        }
+        if left.end_of_epoch_data != right.end_of_epoch_data {
+            this.end_of_epoch_data = Some((
+                left.end_of_epoch_data.clone(),
+                right.end_of_epoch_data.clone(),
+            ));
+        }
+        if left.version_specific_data != right.version_specific_data {
+            this.version_specific_data = Some((
+                left.version_specific_data.clone(),
+                right.version_specific_data.clone(),
+            ));
+        }
+
+        let mut left_iter = left_contents.iter_with_signatures().peekable();
+        let mut right_iter = right_contents.iter_with_signatures().peekable();
+
+        let mut left_diff = Vec::new();
+        let mut right_diff = Vec::new();
+
+        loop {
+            if left_iter.peek().is_none() && right_iter.peek().is_none() {
+                break;
+            }
+            if left_iter.peek().is_none() {
+                let (digests, sigs) = right_iter.next().unwrap();
+                right_diff.push((digests, sigs.to_vec()));
+                continue;
+            }
+            if right_iter.peek().is_none() {
+                let (digests, sigs) = left_iter.next().unwrap();
+                left_diff.push((digests, sigs.to_vec()));
+                continue;
+            }
+
+            let next_left = left_iter.next().unwrap();
```

Comment:
Another nit considering this is a debugging tool.

I think the algorithm here assumes that `right` is a subset of `left`, since we always move to the next left if a match is not found.

say we have X = [a,b,c,d,e] and Y = [a,b,d,e]

`diff(X, Y)` produce
left_diff = [c],
right_diff = []

but `diff(Y, X)` produces
left_diff = [d, e]
right_diff = [c, d, e]

I wonder if we could just upon finding the first mismatch between left and right, print the rest of everything on both sides to make the behavior consistent. Not feeling too strong about it though since this is a debugging tool.

Reply:
or maybe i should just use https://docs.rs/diffs/latest/diffs/myers/fn.diff.html